File: java/src/main/java/com/github/algo/graph/RedundantConnection.java
package com.github.algo.graph;

import java.util.HashMap;
import java.util.Map;

public class RedundantConnection {

    private int findParent(int node, Map<Integer, Integer> parent) {
        while (parent.get(node) != node) {
            parent.put(node, parent.get(node));
            node = parent.get(node);
        }
        return node;
    }

    public int[] findRedundantConnection(int[][] edges) {
        Map<Integer, Integer> parent = new HashMap<>();
        Map<Integer, Integer> rank = new HashMap<>();

        for (int[] edge : edges) {
            parent.putIfAbsent(edge[0], edge[0]);
            parent.putIfAbsent(edge[1], edge[1]);
            rank.putIfAbsent(edge[0], 1);
            rank.putIfAbsent(edge[1], 1);
            int parent1 = findParent(edge[0], parent);
            int parent2 = findParent(edge[1], parent);
            if (parent1 != parent2) {
                if (rank.get(parent1) < rank.get(parent2)) {
                    parent.put(parent1, parent2);
                    rank.put(parent2, rank.get(parent2) + 1);
                } else {
                    parent.put(parent2, parent1);
                    rank.put(parent1, rank.get(parent1) + 1);
                }
            } else {
                return edge;
            }
        }
        return null;
    }

}

File: python/main/graphs/RedundantConnection.py
from typing import List


class Solution:

    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        parent = {}
        rank = {}

        def find(node):
            while parent[node] != node:
                parent[node] = parent[parent[node]]
                node = parent[node]
            return parent[node]

        for s, d in edges:
            if s not in parent:
                parent[s] = s
                rank[s] = 1
            if d not in parent:
                parent[d] = d
                rank[d] = 1

            ps = find(s)
            pd = find(d)

            if ps != pd:
                if rank[ps] > rank[pd]:
                    parent[pd] = ps
                    rank[ps] += 1
                else:
                    parent[ps] = pd
                    rank[pd] += 1
            else:
                return [s, d]
        return -1
        pass


if __name__ == "__main__":
    solution = Solution()
    output = solution.findRedundantConnection([[1, 2], [1, 3], [2, 3]])
    print(output)
